cmd/workload-daemon/main.go

```diff
@@ -0,0 +1,133 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/ingress-gce/cmd/workload-daemon/app"
+	"k8s.io/klog"
+
+	workloadv1a1 "k8s.io/ingress-gce/pkg/apis/workload/v1alpha1"
+	workloadclient "k8s.io/ingress-gce/pkg/experimental/workload/client/clientset/versioned"
+
+	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
+)
+
+const updateInterval time.Duration = 30 * time.Second
+
+func main() {
+	if len(os.Args) < 2 {
+		fmt.Printf("Usage: %v [command] \n", os.Args[0])
+		return
+	}
+	switch os.Args[1] {
+	case "get-credentials":
+		credentials := app.NewCloudVM().Credentials()
+		app.OutputCredentials(credentials)
+		return
+	case "start":
+		vm := app.NewCloudVM()
+		var workload app.WorkloadInfo = vm
+		var helper app.ConnectionHelper = vm
+
+		// Generate KubeConfig and connect to it
+		config := helper.KubeConfig()
+		var clientset workloadclient.Interface
+		clientset, err := workloadclient.NewForConfig(config)
+		if err != nil {
+			klog.Fatalf("unable to connect to the cluster: %+v", err)
+		}
+
+		// Create the workload resource
+		client := clientset.NetworkingV1alpha1().Workloads(corev1.NamespaceDefault)
+		_, err = client.Create(context.Background(), getWorkloadCR(workload), metav1.CreateOptions{})
+		if err != nil {
+			klog.Fatalf("unable to create the workload cr: %+v", err)
+		}
+		klog.V(0).Infof("CR created %s", workload.Name())
+
+		// Update the heartbeat regularly
+		ticker := time.NewTicker(updateInterval)
+		quit := make(chan interface{})
+		sigs := make(chan os.Signal, 1)
+		go updateCR(workload, clientset, ticker, sigs, quit)
+
+		signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
+		<-quit
+
+		err = client.Delete(context.Background(), workload.Name(), metav1.DeleteOptions{})
+		if err != nil {
+			klog.Errorf("unable to delete the workload cr: %+v", err)
+		} else {
+			klog.V(0).Infof("CR deleted")
+		}
+
+		return
+	default:
+		fmt.Printf("Usage: %v [command] \n", os.Args[0])
+		return
+	}
+}
+
+func updateCR(workload app.WorkloadInfo, clientset workloadclient.Interface, ticker *time.Ticker,
+	sigs chan os.Signal, quit chan interface{}) {
+	for {
+		select {
+		case <-ticker.C:
+			patchStr := `[{"op": "replace", "path": "/status/heartbeat", "value": "%s"}]`
+			patch := []byte(fmt.Sprintf(patchStr, time.Now().UTC().Format(time.RFC3339)))
+			vmInstClient := clientset.NetworkingV1alpha1().Workloads(corev1.NamespaceDefault)
+			_, err := vmInstClient.Patch(context.Background(), workload.Name(), types.JSONPatchType, patch, metav1.PatchOptions{})
+			if err != nil {
+				klog.V(0).Infof("CR update failed: %+v", err)
+			} else {
+				klog.V(0).Infof("CR updated %s", workload.Name())
```

Comment:
I'd recommend referencing the resource type here, I think it's workload?
```suggestion
				klog.V(2).Infof("Workload resource updated %s", workload.Name())
```